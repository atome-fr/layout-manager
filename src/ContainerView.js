import React, {Component} from 'react';

// Import component
import View from './View';

/**
 * Class used to manage a container view
 * @class
 */
export default class ContainerView extends Component {

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        let style = {};
        if (this.props.type === 'row'){
            style['height'] = `calc(${this.props.ratio * 100}% - 5px)`;
        } else{
            style['width'] = `calc(${this.props.ratio * 100}% - 5px)`;
        }
        return (
            <div className={this.props.type} style={style}>
                {
                    React.Children.map(this.props.children, child => {
                        if (child.props.visible && child.type === "div") {
                            return <View>
                                {child}
                            </View>
                        }
                        else if (child.type !== "div") {
                            return child
                        }
                    })
                }
            </div>
        );
    }
}