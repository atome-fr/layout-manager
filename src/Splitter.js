

class Splitter extends EventEmmiter{

	constructor(){
		super();
		this.element = $('<div>');

	}
}