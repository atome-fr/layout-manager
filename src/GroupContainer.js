import React, {Component} from 'react';

// Import component
import ContainerView from './ContainerView';
import Splitter from './Splitter';
import ReactDOM from "react-dom";

// Import helpers
import Util from './helpers/util';

/**
 * Class used to manage a group of containers
 * @class
 */
export default class GroupContainer extends Component {

    /**
     * Override the constructor
     * @param props
     */
    constructor(props) {
        super(props);

        this.handleSplitterChange = this.handleSplitterChange.bind(this);
        this._computeSize = this._computeSize.bind(this);

        this.sliderSize = 10;

        this.splitters = [];
        this.children = [];

        if (props.split === 'vertical') {
            this.sizePropName = "width";
            this.offsetPropName = "x";
            this.splitterType = props.split;
            this.containerType = "column";
        }
        else if (props.split === 'horizontal') {
            this.sizePropName = "height";
            this.offsetPropName = "y";
            this.splitterType = props.split;
            this.containerType = "row";
        }
    }

    /**
     * Component lifecycle method
     */
    componentDidMount() {
        this._computeSize(true);
        window.addEventListener("resize", this._computeSize);
    }

    /**
     * Component lifecycle method
     */
    componentDidUpdate() {
        if (this.children.length > 0) {
            this._computeSize();
        }
    }

    /**
     * Component lifecycle method
     */
    componentWillUnmount() {
        window.removeEventListener("resize", this._computeSize);
    }

    /**
     * Handler when the splitter is moved
     * @param splitterOffset The position of the slider
     * @param splitterNode The splitter node
     */
    handleSplitterChange(splitterOffset, splitterNode) {
        const sizeAvailable = this.container[`client${this.sizePropName.charAt(0).toUpperCase() + this.sizePropName.slice(1)}`] - (this.splitters.length * this.sliderSize);
        const ratioOffset = splitterOffset[this.offsetPropName] / sizeAvailable;
        const splitterIndex = Util.searchSplitterIndex(this.splitters, splitterNode);

        this._updateRatios(splitterIndex, ratioOffset);

        this._applyRatios();
    }

    /**
     * Method for update the ratio for the view components
     * @param index The index of the ratio to change
     * @param ratio The new ratio to change
     * @private
     */
    _updateRatios(index, ratio) {
        const newRatios = this.ratios.slice();

        let ratioToAdd = ratio;
        let indexToAdd = index;
        while (ratioToAdd !== 0 && indexToAdd >= 0) {
            if (newRatios[indexToAdd] + ratioToAdd > 0) {
                newRatios[indexToAdd] += ratioToAdd;
                ratioToAdd = 0;
            } else {
                ratioToAdd += newRatios[indexToAdd];
                newRatios[indexToAdd] = 0;
                indexToAdd--;
            }
        }

        let ratioToRem = ratio;
        let indexToRem = index + 1;
        while (ratioToRem !== 0 && indexToRem < this.children.length) {
            if (newRatios[indexToRem] - ratioToRem > 0) {
                newRatios[indexToRem] -= ratioToRem;
                ratioToRem = 0;
            } else {
                ratioToRem -= newRatios[indexToRem];
                newRatios[indexToRem] = 0;
                indexToRem++;
            }
        }

        const sumRatios = newRatios.reduce((acc, val) => acc + val, 0);
        if (sumRatios < 0.99999 || sumRatios > 1.00001) {
            return;
        }

        this.ratios = newRatios;

        this.props.onResize();
    }

    /**
     * Method for apply the ratio to the components
     * @private
     */
    _applyRatios() {
        for (let i = 0; i < this.children.length; i++) {
            ReactDOM.findDOMNode(this.children[i]).style[this.sizePropName] = this.ratios[i] * 100 + '%';
        }
    }

    /**
     * Method for compute the size for each component
     * @private
     */
    _computeSize(isInitialization = false) {

        this.ratios = [1];

        if (this.children.length === 2 && this.props.size) {
            this.ratios = this.props.size;
        }
        else if (this.children.length === 2 && !this.props.size) {
            this.ratios = [0.5, 0.5];
        }

        this._applyRatios();

        if (typeof(isInitialization) !== "boolean" || !isInitialization) {
            document.dispatchEvent(new Event("resizeComponents"));
        }
    }

    /**
     * Method for generate the render
     * @returns {Array}
     * @private
     */
    _generateRender() {
        let toRender = [];

        // Reset array of refs
        this.splitters = [];
        this.children = [];

        let index = 0;

        React.Children.forEach(this.props.children, (child, i) => {
            if ((typeof child.type === "function" && child.type === SplitLayout && Util.childShouldBeDisplay(child)) || child.props.visible) {
                if (index > 0) {
                    toRender.push(this._generateSplitter(index));
                }

                toRender.push(this._generateContainerView(index, child));

                index++;
            }
        });

        return toRender;
    }

    /**
     * Method for generate the splitter component
     * @param index The index of the current child
     * @returns {*}
     * @private
     */
    _generateSplitter(index) {
        return <Splitter ref={splitter => {
            if (splitter) this.splitters.push(splitter);
        }}
                         key={'splitter_group_' + this.props.split + '_' + index}
                         type={this.splitterType}
                         onChange={(offset, node) => this.handleSplitterChange(offset, node)}/>
    }

    /**
     * Method for generate the correct container component
     * @param index The index of the current child
     * @param child The child to render
     * @returns {*}
     * @private
     */
    _generateContainerView(index, child) {
        return <ContainerView ref={col => {
            if (col) this.children.push(col);
        }}
                              key={this.containerType + '_' + index}
                              type={this.containerType}
        >
            {child}
        </ContainerView>
    }

    /**
     * Method for render the components
     * @returns {*}
     */
    render() {
        return (
            <div ref={(el) => this.container = el}
            style={{width:'100%', height:'50%'}}>

                {this._generateRender()}

            </div>
        );
    }
}