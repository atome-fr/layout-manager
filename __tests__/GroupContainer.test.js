import React from 'react';
import ReactDOM from 'react-dom';
import Splitter from '../src/Splitter';
import ContainerView from '../src/ContainerView';
import GroupContainer from '../src/GroupContainer';
import Util from '../src/helpers/util';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('GroupContainer Component', () => {

    let oldVal = null;

    beforeEach(() => {
        oldVal = GroupContainer.prototype.componentDidMount;
        GroupContainer.prototype.componentDidMount = jest.fn();
    });

    afterEach(() => {
        GroupContainer.prototype.componentDidMount = oldVal;
    });

    describe('UI Render Test', () => {

        it('should render with GroupContainer child component', () => {
            const wrapper = shallow(<GroupContainer split="horizontal">
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);
            const tree = toJson(wrapper);

            expect(tree).toMatchSnapshot();
            expect(wrapper.find("div").length).toBe(3);
        });

    });

    describe('Component Method Test', () => {

        it('should call the handleSplitterChange method and call the methods for update the size of the different components', () => {
            const oldVal2 = Util.searchSplitterIndex;
            Util.searchSplitterIndex = jest.fn(() => 0);

            const oldVal3 = GroupContainer.prototype._updateRatios;
            GroupContainer.prototype._updateRatios = jest.fn();

            const oldVal4 = GroupContainer.prototype._applyRatios;
            GroupContainer.prototype._applyRatios = jest.fn();

            const wrapper = shallow(<GroupContainer split="vertical">
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            wrapper.instance().container = {
                clientWidth: 300
            };

            wrapper.instance().handleSplitterChange({x: 1, y: 1}, {});

            expect(Util.searchSplitterIndex).toHaveBeenCalledTimes(1);
            expect(Util.searchSplitterIndex).toHaveBeenCalledWith([], {});
            expect(GroupContainer.prototype._updateRatios).toHaveBeenCalledTimes(1);
            expect(GroupContainer.prototype._updateRatios).toHaveBeenCalledWith(0, (1 / 300));
            expect(GroupContainer.prototype._applyRatios).toHaveBeenCalledTimes(1);

            Util.searchSplitterIndex = oldVal2;
            GroupContainer.prototype._updateRatios = oldVal3;
            GroupContainer.prototype._applyRatios = oldVal4;
        });

        it('should call the _updateRatios method and update the ratio of the container', () => {
            const onResize = jest.fn();

            const wrapper = shallow(<GroupContainer split="vertical" onResize={onResize}>
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            wrapper.instance().ratios = [0.2, 0.2, 0.6];
            wrapper.instance().children = {
                length: 3
            };

            // Should update the ratios
            wrapper.instance()._updateRatios(0, 0.2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);

            // Shouldn't update the ratios
            wrapper.instance()._updateRatios(0, 2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);

            expect(onResize).toHaveBeenCalledTimes(1);
        });

        it('should call the _updateRatios method and update the ratio of the container', () => {
            const onResize = jest.fn();

            const wrapper = shallow(<GroupContainer split="vertical" onResize={onResize}>
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            wrapper.instance().ratios = [0.2, 0.2, 0.6];
            wrapper.instance().children = {
                length: 3
            };

            // Should update the ratios
            wrapper.instance()._updateRatios(0, 0.2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);

            // Shouldn't update the ratios
            wrapper.instance()._updateRatios(0, 2);
            expect(wrapper.instance().ratios).toEqual([0.4, 0, 0.6]);

            expect(onResize).toHaveBeenCalledTimes(1);
        });

        it('should call the _applyRatios method and set the correct size to the children', () => {
            const child1 = {
                style: {
                    width: 0
                }
            };

            const child2 = {
                style: {
                    width: 0
                }
            };

            const oldVal = ReactDOM.findDOMNode;
            ReactDOM.findDOMNode = jest.fn().mockReturnValueOnce(child1).mockReturnValueOnce(child2);

            const wrapper = shallow(<GroupContainer split="vertical">
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            wrapper.instance().ratios = [0.5, 0.5];
            wrapper.instance().children = [child1, child2];

            wrapper.instance()._applyRatios();

            expect(child1.style.width).toBe('50%');
            expect(child2.style.width).toBe('50%');

            ReactDOM.findDOMNode = oldVal;
        });

        it('should call the _computeSize method and set the ratios for the components of the container', () => {
            const oldVal2 = GroupContainer.prototype._applyRatios;
            GroupContainer.prototype._applyRatios = jest.fn();

            const oldVal3 = document.dispatchEvent;
            document.dispatchEvent = jest.fn();

            const wrapper = shallow(<GroupContainer split="vertical" size={null}>
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            wrapper.instance().container = {
                clientWidth: 300
            };
            wrapper.instance().children = {
                length: 2
            };

            wrapper.instance()._computeSize(true);

            expect(wrapper.instance().ratios).toEqual([0.5, 0.5]);
            expect(GroupContainer.prototype._applyRatios).toHaveBeenCalledTimes(1);
            expect(document.dispatchEvent).toHaveBeenCalledTimes(0);

            wrapper.instance()._computeSize();
            expect(document.dispatchEvent).toHaveBeenCalledTimes(1);

            GroupContainer.prototype._applyRatios = oldVal2;
            document.dispatchEvent = oldVal3;
        });

        it('should call the _generateSplitter method and return a SplitterComponent', () => {
            const wrapper = shallow(<GroupContainer split="vertical">
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            const component = wrapper.instance()._generateSplitter(0);

            expect(component.type).toBe(Splitter);
            expect(component.key).toBe('splitter_group_vertical_0');
        });

        it('should call the _generateContainerView method and return a ContainerView Component', () => {
            const wrapper = shallow(<GroupContainer split="vertical">
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            const component = wrapper.instance()._generateContainerView(0, <div>View 1</div>);

            expect(component.type).toBe(ContainerView);
            expect(component.key).toBe('column_0');
        });

        it('should call the _generateRender method and return the correct array of components to render', () => {
            const wrapper = shallow(<GroupContainer split="vertical">
                <div visible={true}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            const toRender = wrapper.instance()._generateRender();

            expect(toRender.length).toBe(3);
            expect(toRender[0].type).toBe(ContainerView);
            expect(toRender[1].type).toBe(Splitter);
            expect(toRender[2].type).toBe(ContainerView);
        });

        it('should call the _generateRender method and return the correct array of components to render', () => {
            const wrapper = shallow(<GroupContainer split="vertical">
                <div visible={false}>View 1</div>
                <div visible={true}>View 2</div>
            </GroupContainer>);

            const toRender = wrapper.instance()._generateRender();

            expect(toRender.length).toBe(1);
            expect(toRender[0].type).toBe(ContainerView);
        });

    });

});