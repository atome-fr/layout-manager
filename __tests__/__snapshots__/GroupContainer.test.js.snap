// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupContainer Component UI Render Test should render with GroupContainer child component 1`] = `
<div
  style={
    Object {
      "height": "50%",
      "width": "100%",
    }
  }
>
  <ContainerView
    key="row_0"
    type="row"
  >
    <div
      visible={true}
    >
      View 1
    </div>
  </ContainerView>
  <Splitter
    key="splitter_group_horizontal_1"
    onChange={[Function]}
    type="horizontal"
  />
  <ContainerView
    key="row_1"
    type="row"
  >
    <div
      visible={true}
    >
      View 2
    </div>
  </ContainerView>
</div>
`;
